Add Phoenix (Deluxe Edition)

ildeniz-s opened the pull request:
Hey, first of all thank you for creating this app, I really like it and don't play the game without it anymore:)!
In this PR, I've added the Deluxe Edition version of the Phoenix card and added some tests. 

I wasn't sure about one rule though. The card also counts as Flame and Weather through a BONUS. So whenever
it is in your hand and its not blanked (by a Flood) it has 3 suits (Beast, Flame and Weather). But how many suits does it
have when its in discard? For example, does Dark Queen get +5 when Phoenix is in discard? My implementation counts it only as a Beast whenever its in discard (so no +5 for Dark Queen) because my understanding is that the BONUS of a card only makes sense as long as it is in the hand of a player (which is the case for ALL other cards in the game). Besides they could have printed this card with triple colors instead of making it a BONUS, if they wanted for it to have triple suits when it is in discard, right? 
Following the same logic, I implemented it so that it can not be duplicated by a Mirage, since I count it only as a Beast when it is not in the hand of a player. What do you think about this approach in general?

I also count it only as a Beast whenever it is copied (by Shapeshifter and Doppelgänger) because these cards do not copy the BONUS of a card.

Apart from adding the card I made a small change in the blanking logic. There is a test case with the description "Bug?" (Warship, War Dirigible, Light Cavalry, Air Elemental, Book of Changes:War Dirigible -> Flood) and it think it was in fact a bug :) 
In this case we have two cards which have a blankedIf method; Warship and War Dirigible. When going over the cards and running their blankedIf methods, the blanking of these cards change depending on the order in which we run them. There are two possibilities:

1) Run blankedIf method of Warship first, then run War Dirigible
    In this case Warship is not blanked because the War Dirigible (which is a Flood) is not blanked yet. Then, War Dirigible becomes blank because of the Air Elemental --> Warship: NOT BLANKED, War Dirigible: BLANKED
2) Run blankedIf method of War Dirigible first, then run Warship
    In this case War Dirigible is blanked because of the Air Elemental, then Warship is also blanked because War Dirigible is blanked --> Warship: BLANKED, War Dirigible: BLANKED

I think the blanking should not depend on the order, thats why I implemented it so that it keeps going over the cards until no new card has been blanked.

Let me know if you find any issues!

Commit:
Until F06 + book of changes

Commit:
finish phoenix interactions

Commit:
fix bugs

jpraet commented on i18n/Messages.properties:
```diff
@@ -141,6 +141,9 @@ FR53.bonus=<b>Doppelgänger</b> may duplicate the name, base strength, suit, and
 FR54.name=Jester
 FR54.bonus=+3 for each other card with an odd base value. <br />OR +50 if entire hand has odd base values.
+FR55.name=Phoenix
+FR55.bonus=Also counts as a <span class="flame">Flame</span> and <span class="weather">Weather</span> card.
```
The remainder of the card text could be added here:
` <br /><b>Phoenix</b> is immune to the <b>Book of Changes</b> and may not BLANK or be BLANKED by any other card.`

ildeniz-s replied:
Done! c7de41d1fba5a4e3ef9c65cbcfb86f48556d6d4a

jpraet commented on js/deck.js:
```diff
@@ -438,7 +449,8 @@ var base = {
     bonusScore: function(hand) {
-      return 5 * (hand.countSuit('land') + hand.countSuit('weather') + hand.countSuit('flood') + hand.countSuit('flame'));
+      return 5 * (hand.countSuit('land') + hand.countSuit('weather') + hand.countSuit('flood') + hand.countSuit('flame'))
+        + hand.containsId(PHOENIX, false) ? 5 : 0;
```
Enchantress = +5 for each Land, Weather, Flood, and Flame.

-> should it not give +10 for Phoenix? As it's both a Weather and a Flame.

ildeniz-s replied:
You are right, for some reason I thought we should count Phoenix only once but your approach makes more sense. Btw the same logic also applies for Blizzard. I fixed them both and added test cases for both of them here 0b17442999a7ff5f6f5e17f8413fe16e2013147a

jpraet commented on js/deck.js:
```diff
@@ -118,7 +118,7 @@ var base = {
       return (card.suit === 'army' && !isArmyClearedFromPenalty(this, hand)) ||
         (card.suit === 'land' && card.name !== 'Mountain') ||
-        (card.suit === 'flame' && card.name !== 'Lightning');
+        (card.suit === 'flame' && card.name !== 'Lightning') || card.id === PHOENIX;
```
I'm not sure about this. If Phoenix's "BLANKED with any Flood" penalty is somehow cleared (e.g. with Beastmaster), should Great Flood then blank Phoenix? Phoenix "may not BLANK or be BLANKED by any other card".
What does "any other card" mean really?

brandonscript replied:
I read this as:

Phoenix is *always* blanked with a flood, unless the flood's penalty is removed.
Other than floods, Phoenix cannot be blanked by any non-flood card, and it cannot contribute to blanking any other card (e.g., if there was a "BLANKED with any FLAME in your hand" card, it would nullify that).

ildeniz-s replied:
My understanding was exactly the same as @brandonscript. That's why I made it so that Great Flood still blanks Phoenix, even though there is a Beastmaster in hand (see the test "Great Flood can still blank a Phoenix since it is a Flood")

jpraet replied:
As there are 2 conflicting texts: "Phoenix cannot be blanked by any other card" vs "Phoenix is blanked with any Flood", I would personally interpret it as: "Phoenix cannot be blanked except by its own penalty". In that case if Phoenix's penalty is cleared, it would not be blanked even by a Flood. But I'm not sure. I'll reach out to my contact for a rules clarification.

jpraet reviewed (commented):
Thanks for this!

The Phoenix card raises quite a lot of questions. There are also 2 versions of it: deluxe version has different card text than the regular version. I wonder if the deluxe version is considered an official errata to the regular version.

I added a couple of review remarks.
Other question: should it be possible to clear Phoenix's penalty with Island? Currently not possible.

Could you extract the change in the blanking logic to a separate PR?

brandonscript commented:
HA! Came here for this. Beat me to it.

Commit:
Add more card text to Phoenix

Commit:
Revert blanking logic fix

Commit:
Island can clear the Penalty of Phoenix

Commit:
Phoenix counts double for Blizzard and Enchantress

ildeniz-s commented:
Thank you for the review!

I also think that the deluxe edition version may be an errata, but regardless, there are probably some people who only own the old version of the card. For them, maybe it would make sense to add that card separately so that we have 2 versions of Phoenix under Beasts and we could also give them different names (eg Phoenix and Phoenix DE). 

Regarding the Island card. I think it should be able to clear the Phoenix's penalty as you said, since penalty clearings happen before blanking. Fixed it here: a35d55df4c5246a96bf6e5b95c731106121cc5a3

I've also reverted the blanking logic fix (and the related test) and will create a separate PR for it.
